webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/ncf/TechniqueReader.scala

```diff
@@ -0,0 +1,85 @@
+package com.normation.rudder.ncf
+
+import better.files.File
+import com.normation.errors.IOResult
+import com.normation.errors.Inconsistency
+import com.normation.rudder.hooks.Cmd
+import com.normation.rudder.hooks.RunNuCommand
+import com.normation.rudder.rest.RestExtractorService
+import net.liftweb.common.EmptyBox
+import net.liftweb.common.Full
+import net.liftweb.json.JsonAST.JArray
+import net.liftweb.json.JsonAST.JObject
+import net.liftweb.json.parse
+import zio.ZIO
+
+class TechniqueReader(
+  restExtractor  : RestExtractorService
+) {
+
+
+  import zio.syntax._
+  import better.files.File._
+  val configuration_repository = root / "var" / "rudder" / "configuration-repository"
+  val methodsFile = configuration_repository / "ncf" / "generic_methods.json"
+  def getAllTechniqueFiles(currentPath : File): List[File]= {
+    val (subdirs, files) = currentPath.children.partition(_.isDirectory)
+    val techniqueFilePath = currentPath / "technique.json"
+    val checkSubdirs = subdirs.flatMap(getAllTechniqueFiles).toList
+    if (techniqueFilePath.exists) {
+      techniqueFilePath :: checkSubdirs
+    } else {
+      checkSubdirs
+    }
+  }
+  def readTechniquesMetadataFile: IOResult[List[Technique]] = {
+    for {
+      methods <- readMethodsMetadataFile
+      techniques <- ZIO.foreach(getAllTechniqueFiles(configuration_repository / "techniques"))(
+        techniqueFile =>
+            restExtractor.extractNcfTechnique(parse(techniqueFile.contentAsString), methods, false) match {
```

Comment:
You need to use `toIO` to correctly suspend the effect in box, else evaluation won't happen when it should. 
Then, you can just use `.chainError` for contextualisation:

```
restExtractor.extractNcfTechnique(parse(techniqueFile.contentAsString), methods, false).toIO
  .chainError("An Error occured while extracting data from techniques ncf API")
```